### PR TITLE
Iteratorsize for partition iterator

### PR DESCRIPTION
**Pull request by mschauer**
Before, collect of a partition could fail because it assumed the existence of a length method via the iteratorsize fall back.

### CODE REVIEW

File: test/functional.jl
Comment (tkelman) [inline test/functional.jl]: presumably this should be `v1[1] == v2[1]` ?

**Comment by mschauer** [comment]
The test failure after adressing @tkelman's spot is unrelated, previously passing test here https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.2015

**Comment by mschauer** [comment]
Personal assessment: this change is straight forward.

**Comment by tkelman** [comment]
Thanks for the bump, I agree this looks fine.

I think all the iterator size business still needs more thorough interface documentation, right?

**Comment by mschauer** [comment]
Yes, I wrote https://gist.github.com/mschauer/c8d8b7eb5b455cb12ddc9bda15695203 but I would like to wait for feedback before writing more docs. See issue #15977

**Comment by mschauer** [comment]
Bump.

File: base/iterator.jl
Comment (rfourquet) [inline base/iterator.jl]: The type parameter can be omitted.

File: base/iterator.jl
Comment (rfourquet) [inline base/iterator.jl]: the argument `P` can be omitted, and when rebasing update to the `where` syntax

File: test/functional.jl
Comment (rfourquet) [inline test/functional.jl]: wrap at 92 (IIRC) chars

**Review by rfourquet — CHANGES_REQUESTED** [review]
LGTM! (modulo the few comments)

**Review by Sacha0 — APPROVED** [review]
lgtm! Thanks @mschauer! :)

(The CI failures appear unrelated.)

**Comment by StefanKarpinski** [comment]
If this is good to go, please merge, @rfourquet since you have some outstanding review comments.

**Comment by mschauer** [comment]
@rfourquet Let me make one amendment first, after #22691 ``partition(flatten([1,2,3,4,5]),2)`` will actually have a length and not trigger anymore the ``SizeUnknown``-code path of collect. One can replace with``partition(flatten([[1,2],[3,4],5]),2)``.

File: base/iterators.jl
Comment (rfourquet) [inline base/iterators.jl]: out of curiosity: you switched to the long-form function syntax because of the `where` syntax?

**Comment by rfourquet** [comment]
Thanks @mschauer !

File: base/iterators.jl
Comment (mschauer) [inline base/iterators.jl]: yes

**Pull request merged**
